package/network/utils/iproute2/patches/191-iplink_allow_to_change_iplink_value.patch

```diff
@@ -0,0 +1,149 @@
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit
+X-Virus-Scanned: clamav-milter 0.100.3 at mail.nic.cz
+X-Virus-Status: Clean
+Sender: netdev-owner@vger.kernel.org
+Precedence: bulk
+List-ID: <netdev.vger.kernel.org>
+X-Mailing-List: netdev@vger.kernel.org
+Archived-At: <https://lore.kernel.org/netdev/20190824024251.4542-5-marek.behun@nic.cz/>
+List-Archive: <https://lore.kernel.org/netdev/>
+List-Post: <mailto:netdev@vger.kernel.org>
+
+Allow to change the interface to which a given interface is linked to.
+This is useful in the case of multi-CPU port DSA, for changing the CPU
+port of a given user port.
+
+Signed-off-by: Marek Behún <marek.behun@nic.cz>
+Cc: David Ahern <dsahern@gmail.com>
+Cc: Stephen Hemminger <stephen@networkplumber.org>
+---
+ ip/iplink.c           | 16 +++++-----------
+ man/man8/ip-link.8.in |  7 +++++++
+ 2 files changed, 12 insertions(+), 11 deletions(-)
+
+diff --git a/ip/iplink.c b/ip/iplink.c
+index 212a0885..d52c0aaf 100644
+--- a/ip/iplink.c
++++ b/ip/iplink.c
+@@ -579,7 +579,6 @@ int iplink_parse(int argc, char **argv, struct iplink_req *req, char **type)
+ {
+ 	char *name = NULL;
+ 	char *dev = NULL;
+-	char *link = NULL;
+ 	int ret, len;
+ 	char abuf[32];
+ 	int qlen = -1;
+@@ -590,6 +589,7 @@ int iplink_parse(int argc, char **argv, struct iplink_req *req, char **type)
+ 	int numrxqueues = -1;
+ 	int link_netnsid = -1;
+ 	int index = 0;
++	int link = -1;
+ 	int group = -1;
+ 	int addr_len = 0;
+ 
+@@ -620,7 +620,10 @@ int iplink_parse(int argc, char **argv, struct iplink_req *req, char **type)
+ 				invarg("Invalid \"index\" value", *argv);
+ 		} else if (matches(*argv, "link") == 0) {
+ 			NEXT_ARG();
+-			link = *argv;
++			link = ll_name_to_index(*argv);
++			if (!link)
++				return nodev(*argv);
++			addattr32(&req->n, sizeof(*req), IFLA_LINK, link);
+ 		} else if (matches(*argv, "address") == 0) {
+ 			NEXT_ARG();
+ 			addr_len = ll_addr_a2n(abuf, sizeof(abuf), *argv);
+@@ -1004,15 +1007,6 @@ int iplink_parse(int argc, char **argv, struct iplink_req *req, char **type)
+ 			exit(-1);
+ 		}
+ 
+-		if (link) {
+-			int ifindex;
+-
+-			ifindex = ll_name_to_index(link);
+-			if (!ifindex)
+-				return nodev(link);
+-			addattr32(&req->n, sizeof(*req), IFLA_LINK, ifindex);
+-		}
+-
+ 		req->i.ifi_index = index;
+ 	}
+ 
+diff --git a/man/man8/ip-link.8.in b/man/man8/ip-link.8.in
+index a8ae72d2..800aed05 100644
+--- a/man/man8/ip-link.8.in
++++ b/man/man8/ip-link.8.in
+@@ -149,6 +149,9 @@ ip-link \- network device configuration
+ .br
+ .RB "[ " nomaster " ]"
+ .br
++.RB "[ " link
++.IR DEVICE " ]"
++.br
+ .RB "[ " vrf
+ .IR NAME " ]"
+ .br
+@@ -2131,6 +2134,10 @@ set master device of the device (enslave device).
+ .BI nomaster
+ unset master device of the device (release device).
+ 
++.TP
++.BI link " DEVICE"
++set device to which this device is linked to.
++
+ .TP
+ .BI addrgenmode " eui64|none|stable_secret|random"
+ set the IPv6 address generation mode
+-- 
+2.21.0
+
+
```

Comment:
can you please give an example how to change cpu-port?

i tried this way on bananapi-r2-board

```
root@bpi-r2:~# ip link set wan nomaster
root@bpi-r2:~# echo $?
0
root@bpi-r2:~# ip link show wan #eth1 still assigned
5: wan@eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether a6:a1:be:38:b8:2e brd ff:ff:ff:ff:ff:ff
root@bpi-r2:~# ip link set wan master eth0
RTNETLINK answers: Operation not supported
```

Reply:
`ip link set wan link eth0`

Reply:
```
root@bpi-r2:~# ip link set wan link eth0
root@bpi-r2:~# ip link show wan
5: wan@eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether a6:a1:be:38:b8:2e brd ff:ff:ff:ff:ff:ff 
```
this is my iproute2 codebase (upstream+this patch)

https://github.com/frank-w/iproute2/commit/22f315e68c0e043ba04d963498f98681488a95b9

compiled with

`make CC="arm-linux-gnueabihf-gcc" LD="arm-linux-gnueabihf-ld" LDFLAGS=-static`

(disabled netem,genl and following in Makefile to fix crosscompile issue, build only "lib ip tc bridge misc")

Reply:
after adding callback to mt7530.c it seems i can set upstream-port, but got oops

```
[   73.167955] ------------[ cut here ]------------                                                                                            
[   73.172571] kernel BUG at net/core/skbuff.c:109!                                                                                            
[   73.177182] Internal error: Oops - BUG: 0 [#1] SMP ARM
...
[   73.686601] [<c0c61834>] (skb_panic) from [<c09dbf5c>] (skb_send_sock_locked+0x0/0x230)                                                     
[   73.694607] [<c09dbf0c>] (skb_push) from [<c0ba72b0>] (mtk_tag_xmit+0x150/0x198)                                                            
[   73.701997]  r5:c4ad7400 r4:c5d64480                                                                                                        
[   73.705572] [<c0ba7160>] (mtk_tag_xmit) from [<c0ba567c>] (dsa_slave_xmit+0xf4/0x1d0)                                                       
[   73.713395]  r5:c4284000 r4:c5d64480                                                                                                        
[   73.716972] [<c0ba5588>] (dsa_slave_xmit) from [<c09fba0c>] (dev_hard_start_xmit+0x120/0x27c)                                               
[   73.725493]  r7:c33c1f00 r6:00000000 r5:c4284000 r4:c5d64480                                                                                
[   73.731152] [<c09fb8ec>] (dev_hard_start_xmit) from [<c09fc4d8>] (__dev_queue_xmit+0x890/0xa70)
```